ruma-signatures/src/functions.rs

```diff
@@ -567,7 +573,9 @@ where
 /// public_key_map.insert("domain".into(), public_key_set);
 ///
 /// // Verify at least one signature for each entity in `public_key_map`.
-/// assert!(verify_event(&public_key_map, &object, &RoomVersionId::Version6).is_ok());
+/// let verification_result = verify_event(&public_key_map, &object, &RoomVersionId::Version6);
+/// assert!(verification_result.is_ok());
+/// assert!(matches!(verification_result.unwrap(), Verified::All));
 /// ```
 pub fn verify_event(
     public_key_map: &PublicKeyMap,
```

Comment:
This `PublicKeyMap` forces to load all the known server keys before calling to this function. I was thinking about adding a trait to be able to provide multiple implementations and, for example, be able to provide an implementation which loads the keys on demand.

What do you think about it?

Reply:
Seems like the kind of situation where you would use a `Stream`, but I'm not sure. Whatever you think works best.

Reply:
Well, as it is right now, `PublicKeyMap` should contain all the public keys for all known servers. What I was thinking on was to have an alternative implementation which lazy loads the public keys on demand (`Stream` does not solve this problem, as far as I know). If I add the trait, I would do it on another PR. 

Thanks anyway for the suggestion!

Reply:
Why does `Stream` not solve this? A stream is basically just an asynchronous (which is just a guess that it'd be useful for homeservers) iterator, and iterators only do work when iterated. When you stop iteration (or polling in the case of a stream) early, any logic that's part of the iterator / stream itself is not invoked again to produce the remaining elements. Of course it somewhat depends on how the stream is constructed.

Reply:
I think that I probably missed in giving enough context. What the `verify_event` needs is random access to one or two keys on the `PublicKeyMap` ( for example, it will need to look up the server keys for `matrix.org` if it needs to authorize an event sent from a user on `matrix.org`). The function does not need to iterate over all the known server keys.

The `Stream` approach could work if the signing keys were selected before hand (for example, select `matrix.org` and `anotherserver.com`) and then use this stream as an input for this function. The unique problem I see with this approach is that it could potentially be misused by, for example, give an empty `Stream` and do not verify any signature.

Reply:
Makes a lot of sense! So your idea was to provide a generic argument that the function can then ask for the keys to a certain homeserver through a trait? Sounds like a good idea!

Reply:
I wonder whether it actually needs a new trait. Could it just be a `FnMut(&ServerName) -> PublicKeySet` / `FnMut(&str) -> PublicKeySet` or sth. along those lines?

Reply:
Yes, that could work as well! I thought on adding a trait just for the semantics.

I do not know if it makes any difference on compilation times :thinking:

Reply:
Not sure how helpful this is but [here](https://gitlab.com/DevinR528/conduit/-/blob/correct-sendtxn/src/server_server.rs#L601-630) is the code for dealing with this so far in conduit.